backend/src/main/java/vn/edu/hust/nmcnpm_20242_n3/controller/SubscriptionController.java

```diff
@@ -0,0 +1,122 @@
+        try {
+            subscriptionService.notifyUsers(bookCopyId);
+            return ResponseEntity.ok("Notifications sent successfully");
+        } catch (Exception e) {
+            return ResponseEntity.status(500).body("Error notifying users: " + e.getMessage());
+        }
+    }
+
+    // API để mượn sách chỉ định và xử lý logic đăng ký nếu sách không khả dụng
+    @PostMapping("/{userId}/borrow")
+    public ResponseEntity<?> borrowBookCopy(@RequestParam int bookCopyId, @PathVariable String userId) {
+        try {
+            subscriptionService.borrowBookCopy(bookCopyId, userId);
+            return ResponseEntity.ok("Borrowing process completed");
+        } catch (Exception e) {
+            return ResponseEntity.status(500).body("Error borrowing book copy: " + e.getMessage());
+        }
+    }
+
+    // API để mượn sách ngẫu nhiên và xử lý logic đăng ký nếu sách không khả dụng
+    @PostMapping("/{userId}/borrow/rand")
+    public ResponseEntity<?> borrowBookCopyRandom(@RequestParam int bookId, @PathVariable String userId) {
+        try {
+            subscriptionService.borrowBookCopyRandom(bookId, userId);
+            return ResponseEntity.ok("Borrowing process completed");
+        } catch (Exception e) {
+            return ResponseEntity.status(500).body("Error borrowing book copy: " + e.getMessage());
+        }
+    }
+
+    // API để hủy đăng ký sau khi mượn sách
+    @DeleteMapping("/cancel-after-borrow/{userId}/{bookCopyId}")
+    public ResponseEntity<?> cancelSubscriptionAfterBorrowing(@PathVariable int bookCopyId, @PathVariable String userId) {
+        try {
+            subscriptionService.cancelSubscriptionAfterBorrowing(bookCopyId, userId);
+            return ResponseEntity.ok("Subscription canceled successfully");
+        } catch (Exception e) {
+            return ResponseEntity.status(500).body("Error canceling subscription: " + e.getMessage());
+        }
+    }
+
+    // API để quản trị viên thông báo thủ công
+    @PostMapping("/notify-all")
+    public ResponseEntity<?> notifyAllUsers() {
+        try {
+            subscriptionService.notifyAllUsers();
+            return ResponseEntity.ok("All users notified successfully");
+        } catch (Exception e) {
+            return ResponseEntity.status(500).body("Error notifying users: " + e.getMessage());
+        }
+    }
+
+    @Scheduled(cron = "0 0 4 * * ?")
+    public void notifyUsersAutomatically() {
+        subscriptionService.notifyAllUsers();
+    }
+
+    @GetMapping("/subscriptions/{userId}")
+    public ResponseEntity<?> getUserSubscriptions(@PathVariable String userId) {
+        try {
+            var subscriptions = subscriptionService.getUserSubscriptions(userId);
+            return ResponseEntity.ok(subscriptions);
+        } catch (Exception e) {
+            return ResponseEntity.status(500).body("Error retrieving subscriptions: " + e.getMessage());
+        }
+    }
```

Comment:
Okay this will create an endless reference in the response. Should we consider making a `SubscriptionDTO`?

You can check out the same incident here for more info: https://github.com/Sukkaito/IT3180_20242_N3/pull/25#discussion_r2088318053